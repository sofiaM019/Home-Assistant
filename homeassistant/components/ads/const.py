"""Support for Automation Device Specification (ADS)."""

from __future__ import annotations

from enum import StrEnum
from typing import TYPE_CHECKING

from homeassistant.util.hass_dict import HassKey

if TYPE_CHECKING:
    from .hub import AdsHub

DOMAIN = "ads"

DATA_ADS: HassKey[AdsHub] = HassKey(DOMAIN)
STATE_KEY_STATE = "state"

SERVICE_WRITE_DATA_BY_NAME = "write_data_by_name"

CONF_ADS_VAR = "adsvar"
CONF_ADS_FACTOR = "factor"
CONF_ADS_TYPE = "adstype"
CONF_ADS_VALUE = "value"


class AdsType(StrEnum):
    """Supported Types."""

    BOOL = "bool"
    BYTE = "byte"
    INT = "int"
    UINT = "uint"
    SINT = "sint"
    USINT = "usint"
    DINT = "dint"
    UDINT = "udint"
    WORD = "word"
    DWORD = "dword"
    LREAL = "lreal"
    REAL = "real"
    STRING = "string"
    TIME = "time"
    DATE = "date"
    DATE_AND_TIME = "dt"
    TOD = "tod"
