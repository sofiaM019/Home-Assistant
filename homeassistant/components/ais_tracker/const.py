"""AIS tracker constants."""

DOMAIN = "ais_tracker"

CONF_MMSIS = "mmsis"
DEFAULT_PORT = 12346
