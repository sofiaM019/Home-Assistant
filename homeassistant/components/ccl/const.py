"""Constants for the CCL Electronics integration."""

DOMAIN = "ccl"
