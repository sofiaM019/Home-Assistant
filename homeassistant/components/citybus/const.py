"""CityBus constants"""

DOMAIN = "citybus"

CONF_ROUTE = "route"
CONF_DIRECTION = "direction"
CONF_STOP = "stop"
