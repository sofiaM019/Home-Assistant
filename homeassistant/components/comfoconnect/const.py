"""Constants for the ComfoConnect integration."""

DOMAIN = "comfoconnect"

CONF_USER_AGENT = "user_agent"
DEFAULT_PIN = 0
DEFAULT_TOKEN = "00000000000000000000000000000001"
DEFAULT_USER_AGENT = "Home Assistant"
