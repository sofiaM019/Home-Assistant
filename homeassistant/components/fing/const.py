"""Const for the Fing integration."""

AGENT_IP = "agent_ip"
AGENT_PORT = "agent_port"
AGENT_KEY = "agent_key"
DOMAIN = "fing"
