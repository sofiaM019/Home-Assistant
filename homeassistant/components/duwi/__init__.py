"""Support for Duwi Smart devices."""

from typing import Any, NamedTuple

from duwi_smarthome_sdk.base_api import CustomerApi
from duwi_smarthome_sdk.device_scene_models import CustomerDevice, CustomerScene
from duwi_smarthome_sdk.manager import Manager, SharingDeviceListener

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import __version__
from homeassistant.core import HomeAssistant, callback
from homeassistant.exceptions import ConfigEntryAuthFailed
from homeassistant.helpers import device_registry as dr
from homeassistant.helpers.dispatcher import dispatcher_send

from .const import (
    _LOGGER,
    ACCESS_TOKEN,
    ADDRESS,
    APP_KEY,
    APP_SECRET,
    CLIENT_MODEL,
    CLIENT_VERSION,
    DOMAIN,
    DUWI_DISCOVERY_NEW,
    DUWI_HA_SIGNAL_UPDATE_ENTITY,
    DUWI_SCENE_UPDATE,
    HOUSE_KEY,
    HOUSE_NAME,
    HOUSE_NO,
    PASSWORD,
    PHONE,
    REFRESH_TOKEN,
    SUPPORTED_PLATFORMS,
    WS_ADDRESS,
)

type DuwiConfigEntry = ConfigEntry[HomeAssistantDuwiData]


class HomeAssistantDuwiData(NamedTuple):
    """Store Duwi data in the Home Assistant data object."""

    manager: Manager
    listener: SharingDeviceListener


async def async_setup_entry(hass: HomeAssistant, entry: DuwiConfigEntry) -> bool:
    """Set up a config entry for the Duwi Smart Devices integration."""

    hass.data.setdefault(DOMAIN, {})

    manager: Manager = Manager(
        customer_api=CustomerApi(
            address=entry.data[ADDRESS],
            ws_address=entry.data[WS_ADDRESS],
            app_key=entry.data[APP_KEY],
            app_secret=entry.data[APP_SECRET],
            house_no=entry.data[HOUSE_NO],
            house_name=entry.data[HOUSE_NAME],
            access_token=entry.data[ACCESS_TOKEN],
            refresh_token=entry.data[REFRESH_TOKEN],
            client_version=CLIENT_VERSION,
            client_model=CLIENT_MODEL,
            app_version=__version__,
        ),
        house_key=entry.data.get(HOUSE_KEY),
        token_refresh_callback=update_token,
    )

    # Fetch devices
    await manager.init_manager()
    await manager.update_device_cache()
    listener = DeviceListener(hass, manager)
    manager.add_device_listener(listener)

    entry.runtime_data = HomeAssistantDuwiData(manager=manager, listener=listener)

    # Clean up inappropriate devices
    await cleanup_device_registry(hass, entry)

    # Start global WebSocket listener
    await hass.async_create_task(manager.ws.reconnect())

    hass.loop.create_task(manager.ws.listen())
    hass.loop.create_task(manager.ws.keep_alive())

    await hass.config_entries.async_forward_entry_setups(entry, SUPPORTED_PLATFORMS)

    return True


async def compare_manager(old_manager: Manager, new_manager: Manager) -> list:
    """Compare old and new manager to find devices with different room or type."""

    ids = []

    # Compare devices in both managers
    for dev_id in old_manager.device_map:
        d1 = old_manager.device_map.get(dev_id)
        d2 = new_manager.device_map.get(dev_id)
        if not d2:
            # Device not found in new manager
            continue

        # Compare room and device type; if different, add device ID to the list
        if d1.room_no != d2.room_no or d1.device_sub_type_no != d2.device_sub_type_no:
            ids.append(dev_id)

    return ids


async def cleanup_device_registry(hass: HomeAssistant, entry: DuwiConfigEntry) -> None:
    """Remove deleted device registry entry if there are no remaining entities."""

    device_registry = dr.async_get(hass)

    for dev_id, device_entry in list(device_registry.devices.items()):
        for item in device_entry.identifiers:
            if item[0] == DOMAIN and (
                item[1] not in entry.runtime_data.manager.device_map
            ):
                device_registry.async_remove_device(dev_id)


async def async_unload_entry(hass: HomeAssistant, entry: DuwiConfigEntry) -> bool:
    """Unload the Duwi platforms associated with the provided config entry."""
    if unload_ok := await hass.config_entries.async_unload_platforms(
        entry, SUPPORTED_PLATFORMS
    ):
        await entry.runtime_data.manager.unload(True)

    return unload_ok


class DeviceListener(SharingDeviceListener):
    """Device Update Listener."""

    def __init__(self, hass: HomeAssistant, manager: Manager) -> None:
        """Initialize the DeviceListener."""
        self.hass = hass
        self.manager = manager

    def update_scene(self, scene: CustomerScene) -> None:
        """Notify about scene update."""
        dispatcher_send(self.hass, f"{DUWI_SCENE_UPDATE}_{scene.scene_no}")

    def update_device(self, device: CustomerDevice) -> None:
        """Update device status notification."""
        dispatcher_send(self.hass, f"{DUWI_HA_SIGNAL_UPDATE_ENTITY}_{device.device_no}")

    def add_device(self, device: CustomerDevice) -> None:
        """Handle a new device being added."""
        # Ensure the device isn't present stale
        self.async_remove_device(device.device_no)
        dispatcher_send(self.hass, DUWI_DISCOVERY_NEW, [device.device_no])

    def remove_device(self, device_no: str) -> None:
        """Schedule removal of a device."""
        self.hass.add_job(self.async_remove_device, device_no)

    @callback
    def async_remove_device(self, device_no: str) -> None:
        """Remove device from Home Assistant."""
        device_registry = dr.async_get(self.hass)
        _LOGGER.info("Removing device_no %s", device_no)

        device_entry = device_registry.async_get_device(
            identifiers={(DOMAIN, device_no)}
        )

        if device_entry is not None:
            device_registry.async_remove_device(device_entry.id)


@callback
def update_token(
    self, is_refresh: bool, token_info: dict[str, Any] | None = None
) -> None:
    """Update token info in the config entry."""
    _LOGGER.info("Updating token: %s %s", is_refresh, token_info)

    if is_refresh:
        # Prepare data for updating the config entry
        data = {
            PHONE: self.entry.data[PHONE] if self.entry.data is not None else None,
            PASSWORD: (
                self.entry.data[PASSWORD] if self.entry.data is not None else None
            ),
            HOUSE_KEY: (
                self.entry.data[HOUSE_KEY] if self.entry.data is not None else None
            ),
            ADDRESS: (
                self.entry.data[ADDRESS] if self.entry.data is not None else None
            ),
            WS_ADDRESS: (
                self.entry.data[WS_ADDRESS] if self.entry.data is not None else None
            ),
            APP_KEY: (
                self.entry.data[APP_KEY] if self.entry.data is not None else None
            ),
            APP_SECRET: (
                self.entry.data[APP_SECRET] if self.entry.data is not None else None
            ),
            HOUSE_NO: (
                self.entry.data[HOUSE_NO] if self.entry.data is not None else None
            ),
            ACCESS_TOKEN: (
                token_info[ACCESS_TOKEN] if token_info is not None else None
            ),
            REFRESH_TOKEN: (
                token_info[REFRESH_TOKEN] if token_info is not None else None
            ),
        }
    else:
        raise ConfigEntryAuthFailed(
            "Failed to refresh token, please try reloading the integration or re-adding it."
        )

    @callback
    def async_update_entry() -> None:
        """Update the configuration entry with new token data."""
        self.hass.config_entries.async_update_entry(self.entry, data=data)

    # Schedule the asynchronous update job
    self.hass.add_job(async_update_entry)
