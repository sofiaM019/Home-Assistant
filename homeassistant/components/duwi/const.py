"""Constants for the DUWI integration."""

from enum import StrEnum
import logging

from homeassistant.const import Platform

_LOGGER = logging.getLogger(__package__)

MANUFACTURER = "duwi"

DOMAIN = "duwi"

APP_VERSION = "0.1.1"

CLIENT_VERSION = "0.1.1"

CLIENT_MODEL = "homeassistant"

DUWI_DISCOVERY_NEW = "duwi_discovery_new"
DUWI_SCENE_UPDATE = "duwi_scene_update"
DUWI_HA_SIGNAL_UPDATE_ENTITY = "duwi_entry_update"
DUWI_HA_ACCESS_TOKEN = "duwi_access_token"

# API keys
CLIENT = "client"
ADDRESS = "address"
WS_ADDRESS = "ws_address"
APP_KEY = "app_key"
APP_SECRET = "app_secret"
ACCESS_TOKEN = "access_token"
REFRESH_TOKEN = "refresh_token"
PHONE = "phone"
PASSWORD = "password"
HOUSE_NO = "house_no"
HOUSE_NAME = "house_name"
HOUSE_KEY = "house_key"
CONF_TOKEN_INFO = "token_info"

# List of platforms that support config entry
SUPPORTED_PLATFORMS = [
    Platform.SWITCH,
]


class DPCode(StrEnum):
    """Data Point Codes used by Duwi."""

    # button
    BUTTON = "button"

    # switch
    SWITCH = "switch"

    # light
    ON_OFF = "on_off"
    LIGHT = "light"
    COLOR_TEMP = "color_temp"
    COLOR = "color"
    RGB = "rgb"
    RGBW = "rgbw"
    RGBCW = "rgbcw"

    # cover
    ROLLER = "roller"
    SHUTTER = "shutter"
    SHUTTER_2 = "shutter-2"

    # music
    HUA_ERSI_MUSIC = "hua_ersi_music"
    XIANG_WANG_MUSIC_S7_MINI_3S = "xiangwang_music_s7_mini_3s"
    XIANG_WANG_MUSIC_S8 = "xiangwang_music_s8"
    SHENG_BI_KE_MUSIC = "sheng_bike_music"
    BO_SHENG_MUSIC = "bo_sheng_music"
    YOU_DA_MUSIC = "youda_music"

    # sensor
    TEMP = "temp"
    HUMIDITY = "humidity"
    HCHO = "hcho"
    PM25 = "pm25"
    CO2 = "co2"
    BRIGHT = "bright"
    IQA = "iaq"
    HUMAN = "human"
    TRIGGER = "trigger"
    MOISTURE = "moisture"
    SMOKE = "smoke"
    MOVING = "moving"
    GAS = "gas"
    DOOR = "door"
    OPENING = "opening"
    CO = "co"
    TVOC = "tvoc"
    PM10 = "pm10"

    # climate
    AC = "ac"
    FA = "fa"
    FH = "fh"
    HP = "hp"
    TC = "tc"
