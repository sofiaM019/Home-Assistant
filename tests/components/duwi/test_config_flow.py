"""Define integration configuration flow tests for the Duwi integration."""

from __future__ import annotations

import pytest
from syrupy.assertion import SnapshotAssertion

from homeassistant.components.duwi.const import DOMAIN
from homeassistant.core import HomeAssistant
from homeassistant.data_entry_flow import FlowResultType


@pytest.mark.usefixtures("mock_duwi_login_user", "mock_duwi_login_and_fetch_house")
async def test_user_flow_success(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
) -> None:
    """Test the configuration flow: successful path.

    This simulates a user going through the configuration flow, including
    providing correct credentials and selecting a house, leading to a successful
    creation of a config entry.
    """
    # Start by initializing the flow
    result = await hass.config_entries.flow.async_init(
        DOMAIN,
        context={"source": "user"},
    )
    # Ensure we're being prompted for initial information
    assert result.get("type") == FlowResultType.FORM
    assert result.get("step_id") == "user"
    snapshot.assert_match(result)

    # Simulate submitting the app credentials step
    result2 = await hass.config_entries.flow.async_configure(
        result["flow_id"],
        user_input={
            "app_key": "correct_app_key",
            "app_secret": "correct_app_secret",
            "phone": "correct_phone",
            "password": "correct_password",
        },
    )

    # Move to selecting a house
    assert result2.get("type") == FlowResultType.FORM
    assert result2.get("step_id") == "select_house"
    snapshot.assert_match(result2)

    # Finally, simulate selecting a house
    result3 = await hass.config_entries.flow.async_configure(
        result["flow_id"],
        user_input={"house_no": "mocked_house_no1"},
    )
    # The flow should now complete with a CREATE_ENTRY result
    assert result3.get("type") == FlowResultType.CREATE_ENTRY
    snapshot.assert_match(result3)


@pytest.mark.usefixtures("mock_duwi_login_user", "mock_duwi_login_and_fetch_house")
async def test_user_flow_failed_by_phone_auth(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
) -> None:
    """Test the configuration flow: successful path.

    This simulates a user going through the configuration flow, including
    providing correct credentials and selecting a house, leading to a successful
    creation of a config entry.
    """
    # Start by initializing the flow
    result = await hass.config_entries.flow.async_init(
        DOMAIN,
        context={"source": "user"},
    )
    # Ensure we're being prompted for initial information
    assert result.get("type") == FlowResultType.FORM
    assert result.get("step_id") == "user"

    # Simulate submitting the app credentials step
    result2 = await hass.config_entries.flow.async_configure(
        result["flow_id"],
        user_input={
            "app_key": "correct_app_key",
            "app_secret": "correct_app_secret",
            "phone": "error_phone",
            "password": "correct_password",
        },
    )

    # Move to selecting a house
    assert result2.get("type") == FlowResultType.FORM
    assert result2.get("step_id") == "user"
    snapshot.assert_match(result2)


@pytest.mark.usefixtures("mock_duwi_login_user", "mock_duwi_fetch_house_info_error")
async def test_fetch_house_info_error(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
) -> None:
    """Test the configuration flow when fetch_house_info fails."""

    result = await hass.config_entries.flow.async_init(
        DOMAIN,
        context={"source": "user"},
    )
    assert result.get("type") == FlowResultType.FORM
    assert result.get("step_id") == "user"

    result2 = await hass.config_entries.flow.async_configure(
        result["flow_id"],
        user_input={
            "app_key": "correct_app_key",
            "app_secret": "correct_app_secret",
            "phone": "correct_phone",
            "password": "correct_password",
        },
    )

    # Assert fetch_house_info error is caught
    assert result2.get("type") == FlowResultType.FORM
    assert result2.get("errors") == {"base": "fetch_house_info_error"}
    snapshot.assert_match(result2)


@pytest.mark.usefixtures("mock_duwi_login_user", "mock_duwi_no_houses_found")
async def test_auth_success_but_no_houses_found(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
) -> None:
    """Test the configuration flow when authentication succeeds but no houses are found."""

    # Initialize the configuration flow
    result = await hass.config_entries.flow.async_init(
        DOMAIN,
        context={"source": "user"},
    )
    assert result.get("type") == FlowResultType.FORM
    assert result.get("step_id") == "user"

    # Simulate user input with correct credentials
    result2 = await hass.config_entries.flow.async_configure(
        result["flow_id"],
        user_input={
            "app_key": "correct_app_key",
            "app_secret": "correct_app_secret",
            "phone": "correct_phone",
            "password": "correct_password",
        },
    )

    # Check that the form is returned with the correct error when no houses are found
    assert result2.get("type") == FlowResultType.FORM
    assert result2.get("step_id") == "select_house"
    snapshot.assert_match(result2)


@pytest.mark.usefixtures("mock_duwi_login_invalid_auth")
async def test_invalid_auth_error(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
) -> None:
    """Test the configuration flow with invalid authentication."""

    result = await hass.config_entries.flow.async_init(
        DOMAIN,
        context={"source": "user"},
    )
    assert result.get("type") == FlowResultType.FORM
    assert result.get("step_id") == "user"

    result2 = await hass.config_entries.flow.async_configure(
        result["flow_id"],
        user_input={
            "app_key": "correct_app_key",
            "app_secret": "correct_app_secret",
            "phone": "correct_phone",
            "password": "wrong_password",
        },
    )

    # Assert invalid authentication error
    assert result2.get("type") == FlowResultType.FORM
    assert result2.get("errors") == {"base": "invalid_auth"}
    snapshot.assert_match(result2)


@pytest.mark.usefixtures("mock_duwi_login_sys_error")
async def test_invalid_sys_error(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
) -> None:
    """Test the configuration flow with invalid authentication."""

    result = await hass.config_entries.flow.async_init(
        DOMAIN,
        context={"source": "user"},
    )
    assert result.get("type") == FlowResultType.FORM
    assert result.get("step_id") == "user"

    result2 = await hass.config_entries.flow.async_configure(
        result["flow_id"],
        user_input={
            "app_key": "correct_app_key",
            "app_secret": "correct_app_secret",
            "phone": "correct_phone",
            "password": "wrong_password",
        },
    )

    # Assert invalid authentication error
    assert result2.get("type") == FlowResultType.FORM
    assert result2.get("errors") == {"base": "sys_error"}
    snapshot.assert_match(result2)
