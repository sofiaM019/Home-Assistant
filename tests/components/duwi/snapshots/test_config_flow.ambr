# serializer version: 1
# name: test_auth_error
  FlowResultSnapshot({
    'data_schema': list([
      dict({
        'name': 'app_key',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'app_secret',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'phone',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'password',
        'required': True,
        'type': 'string',
      }),
    ]),
    'description_placeholders': dict({
      'code': 'AUTH_ERROR',
    }),
    'errors': dict({
      'base': 'auth_error',
    }),
    'flow_id': <ANY>,
    'handler': 'duwi',
    'last_step': None,
    'preview': None,
    'step_id': 'user',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_auth_success_but_no_houses_found
  FlowResultSnapshot({
    'data_schema': list([
      dict({
        'name': 'house_no',
        'options': list([
        ]),
        'required': True,
        'type': 'select',
      }),
    ]),
    'description_placeholders': dict({
    }),
    'errors': dict({
      'base': 'no_houses_found_error',
    }),
    'flow_id': <ANY>,
    'handler': 'duwi',
    'last_step': None,
    'preview': None,
    'step_id': 'select_house',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_fetch_house_info_error
  FlowResultSnapshot({
    'data_schema': list([
      dict({
        'name': 'app_key',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'app_secret',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'phone',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'password',
        'required': True,
        'type': 'string',
      }),
    ]),
    'description_placeholders': dict({
      'code': '10001',
    }),
    'errors': dict({
      'base': 'fetch_house_info_error',
    }),
    'flow_id': <ANY>,
    'handler': 'duwi',
    'last_step': None,
    'preview': None,
    'step_id': 'user',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_invalid_auth_error
  FlowResultSnapshot({
    'data_schema': list([
      dict({
        'name': 'app_key',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'app_secret',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'phone',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'password',
        'required': True,
        'type': 'string',
      }),
    ]),
    'description_placeholders': dict({
      'code': '11000',
    }),
    'errors': dict({
      'base': 'invalid_auth',
    }),
    'flow_id': <ANY>,
    'handler': 'duwi',
    'last_step': None,
    'preview': None,
    'step_id': 'user',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_invalid_sys_error
  FlowResultSnapshot({
    'data_schema': list([
      dict({
        'name': 'app_key',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'app_secret',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'phone',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'password',
        'required': True,
        'type': 'string',
      }),
    ]),
    'description_placeholders': dict({
      'code': '10001',
    }),
    'errors': dict({
      'base': 'sys_error',
    }),
    'flow_id': <ANY>,
    'handler': 'duwi',
    'last_step': None,
    'preview': None,
    'step_id': 'user',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_no_houses_found_after_auth
  FlowResultSnapshot({
    'data_schema': list([
      dict({
        'name': 'house_no',
        'options': list([
        ]),
        'required': True,
        'type': 'select',
      }),
    ]),
    'description_placeholders': dict({
    }),
    'errors': dict({
      'base': 'no_houses_found_error',
    }),
    'flow_id': <ANY>,
    'handler': 'duwi',
    'last_step': None,
    'preview': None,
    'step_id': 'select_house',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_sys_error
  FlowResultSnapshot({
    'data_schema': list([
      dict({
        'name': 'app_key',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'app_secret',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'phone',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'password',
        'required': True,
        'type': 'string',
      }),
    ]),
    'description_placeholders': dict({
      'code': '10001',
    }),
    'errors': dict({
      'base': 'sys_error',
    }),
    'flow_id': <ANY>,
    'handler': 'duwi',
    'last_step': None,
    'preview': None,
    'step_id': 'user',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_user_flow_failed_by_phone_auth
  FlowResultSnapshot({
    'data_schema': list([
      dict({
        'name': 'app_key',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'app_secret',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'phone',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'password',
        'required': True,
        'type': 'string',
      }),
    ]),
    'description_placeholders': dict({
      'code': '11000',
    }),
    'errors': dict({
      'base': 'invalid_auth',
    }),
    'flow_id': <ANY>,
    'handler': 'duwi',
    'last_step': None,
    'preview': None,
    'step_id': 'user',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_user_flow_success
  FlowResultSnapshot({
    'data_schema': list([
      dict({
        'name': 'app_key',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'app_secret',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'phone',
        'required': True,
        'type': 'string',
      }),
      dict({
        'name': 'password',
        'required': True,
        'type': 'string',
      }),
    ]),
    'description_placeholders': dict({
    }),
    'errors': dict({
    }),
    'flow_id': <ANY>,
    'handler': 'duwi',
    'last_step': None,
    'preview': None,
    'step_id': 'user',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_user_flow_success.1
  FlowResultSnapshot({
    'data_schema': list([
      dict({
        'name': 'house_no',
        'options': list([
          tuple(
            'mocked_house_no1',
            'mocked_house_name1',
          ),
          tuple(
            'mocked_house_no2',
            'mocked_house_name2',
          ),
        ]),
        'required': True,
        'type': 'select',
      }),
    ]),
    'description_placeholders': dict({
    }),
    'errors': dict({
    }),
    'flow_id': <ANY>,
    'handler': 'duwi',
    'last_step': None,
    'preview': None,
    'step_id': 'select_house',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_user_flow_success.2
  FlowResultSnapshot({
    'context': dict({
      'source': 'user',
    }),
    'data': dict({
      'access_token': None,
      'address': 'https://openapi.duwi.com.cn/homeApi/v1',
      'app_key': 'correct_app_key',
      'app_secret': 'correct_app_secret',
      'house_key': 'mocked_lan_secret_key1',
      'house_name': 'mocked_house_name1',
      'house_no': 'mocked_house_no1',
      'password': 'correct_password',
      'phone': 'correct_phone',
      'refresh_token': None,
      'ws_address': 'wss://openws.duwi.com.cn:32013',
    }),
    'description': None,
    'description_placeholders': None,
    'flow_id': <ANY>,
    'handler': 'duwi',
    'minor_version': 1,
    'options': dict({
    }),
    'result': ConfigEntrySnapshot({
      'data': dict({
        'access_token': None,
        'address': 'https://openapi.duwi.com.cn/homeApi/v1',
        'app_key': 'correct_app_key',
        'app_secret': 'correct_app_secret',
        'house_key': 'mocked_lan_secret_key1',
        'house_name': 'mocked_house_name1',
        'house_no': 'mocked_house_no1',
        'password': 'correct_password',
        'phone': 'correct_phone',
        'refresh_token': None,
        'ws_address': 'wss://openws.duwi.com.cn:32013',
      }),
      'disabled_by': None,
      'domain': 'duwi',
      'entry_id': <ANY>,
      'minor_version': 1,
      'options': dict({
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'user',
      'title': 'mocked_house_name1',
      'unique_id': None,
      'version': 1,
    }),
    'title': 'mocked_house_name1',
    'type': <FlowResultType.CREATE_ENTRY: 'create_entry'>,
    'version': 1,
  })
# ---
